Fix for whip/pike attack invisible but revealed enemies

brturn opened the pull request:
As theorized by @slow-burn on brogue discord, the pike, whip, and flail do not work against invisible but revealed enemies (eg. telepathy).

- Pike & spear work
- Flail works
- Axe works
- Rapier works
- Whip works against invisible enemies
- Whip cannot attack submerged enemies

brturn committed:
Fix for whip/pike attack invisible but revealed enemies

brturn committed:
ChangeLog

brturn committed:
Centralizing hit check logic

brturn committed:
Whip cannot strike submerged enemies

tmewett commented on src/brogue/Monsters.c:
```diff
@@ -230,6 +230,30 @@ boolean canDirectlySeeMonster(creature *monst) {
+
+    // If the player is the attacker, also check vision and telepathy
+    if (attacker == &player) {
+        locationKnown = (locationKnown && playerCanSee(defender->loc.x, defender->loc.y))
+                        || monsterRevealed(defender);
```
This is equiv to canSeeMonster, right? Thoughts on using that?

tmewett replied:
It's not equivalent actually - it's grouped differently. It's equivalent to `canSee || revealed` (which does mean location known) because the other `|| revealed` in `canSee` is made redundant by the outer one.

tmewett reviewed (commented):
Thanks!

Thought this was making hitlist inclusion and attack triggering use the same code, but I see that's only for the flail where that logic actually should be the same.

Just a question.

zenzombie commented on src/brogue/Items.c:
```diff
@@ -3460,15 +3460,13 @@ void getImpactLoc(pos *returnLoc, const pos originLoc, const pos targetLoc,
     pos coords[DCOLS + 1];
     short i, n;
     creature *monst;
+    creature *orig = monsterAtLoc(originLoc.x, originLoc.y);
```
Perhaps *attacker or *caster instead of *orig ?

nstoddard commented on changes/issue-516.md:
```diff
@@ -1,2 +1,2 @@
 -
-  The rapier lunge attack works against invisible enemies when you are telepathic.
+  The whip, spear, and rapier now work against invisible enemies when you are telepathic.
```
Maybe also mention that whips no longer work against submerged enemies?

tmewett commented on src/brogue/Items.c:
```diff
@@ -3460,14 +3460,15 @@ void getImpactLoc(pos *returnLoc, const pos originLoc, const pos targetLoc,
         monst = monsterAtLoc(coords[i].x, coords[i].y);
-        if (monst
-            && !monsterIsHidden(monst, monsterAtLoc(originLoc.x, originLoc.y))
+        if (canAttack(orig, monst)
```
This is wrong from the POV of getImpactLoc. Bolts do not pass through creatures if the caster is unable to attack them.

tmewett commented on src/brogue/Monsters.c:
```diff
@@ -230,6 +230,30 @@ boolean canDirectlySeeMonster(creature *monst) {
     return false;
 }
 
+// Centralize the logic of whether an attacker can/will attack the defender
+boolean canAttack( creature *attacker, creature *defender ) {
```
We should decide what this means precisely. For example, this check includes that the defender's location is known, but it's obviously possible to attack completely invisible enemies.

It probably means something more like "does the attack know it will attack defender if it moves into its cell?"

tmewett reviewed (changes requested):
We should name and document canAttack properly before merging

tmewett committed:
Refactor canAttack's locationKnown

The player case was equivalent to `canSee || revealed` because the inner
`|| revealed` in `canSee` is made redundant by the outer one.

brturn commented:
Closing in favor of #686, which is much cleaner and can be applied universally.
